presto-hive/src/main/java/com/facebook/presto/hive/SortingFileWriter.java

```diff
@@ -0,0 +1,207 @@
+
+    private final FileSystem fileSystem;
+    private final Path tempFilePrefix;
+    private final int maxTempFiles;
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final HiveFileWriter outputWriter;
+    private final SortBuffer sortBuffer;
+    private final List<Path> tempFiles = new ArrayList<>();
+
+    public SortingFileWriter(
+            FileSystem fileSystem,
+            Path tempFilePrefix,
+            HiveFileWriter outputWriter,
+            DataSize maxMemory,
+            int maxTempFiles,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders,
+            PageSorter pageSorter)
+    {
+        checkArgument(maxTempFiles > 0, "maxTempFiles must be greater than zero");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        this.tempFilePrefix = requireNonNull(tempFilePrefix, "tempFilePrefix is null");
+        this.maxTempFiles = maxTempFiles;
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(requireNonNull(sortFields, "sortFields is null"));
+        this.sortOrders = ImmutableList.copyOf(requireNonNull(sortOrders, "sortOrders is null"));
+        this.outputWriter = requireNonNull(outputWriter, "outputWriter is null");
+        this.sortBuffer = new SortBuffer(maxMemory, types, sortFields, sortOrders, pageSorter);
+    }
+
+    @Override
+    public long getWrittenBytes()
+    {
+        return outputWriter.getWrittenBytes();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE + sortBuffer.getRetainedBytes();
+    }
+
+    @Override
+    public void appendRows(Page page)
+    {
+        if (tempFiles.size() == maxTempFiles) {
+            throw new PrestoException(HIVE_TOO_MANY_BUCKET_SORT_FILES, "Too many temporary files for sorted bucket writer");
+        }
+
+        if (!sortBuffer.canAdd(page)) {
+            flushBuffer();
+        }
+        sortBuffer.add(page);
+    }
+
+    @Override
+    public void commit()
+    {
+        if (!sortBuffer.isEmpty()) {
+            flushBuffer();
+        }
+
+        try {
+            writeSorted();
+            outputWriter.commit();
+
+            for (Path file : tempFiles) {
+                if (!fileSystem.delete(file, false)) {
+                    throw new IOException("Failed to delete temporary file: " + file);
+                }
+            }
+        }
+        catch (IOException | UncheckedIOException e) {
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error committing write to Hive", e);
+        }
+    }
+
+    @Override
+    public void rollback()
+    {
+        for (Path file : tempFiles) {
+            try {
+                if (fileSystem.exists(file) && !fileSystem.delete(file, false)) {
```

Comment:
If you did this in the reverse order, you would prevent races... so blindly delete and then check if it still exists.